### ISSUE TITLE
LSTM policy

### ISSUE THREAD

**Issue by nadavbh12** [ikostrikov/pytorch-a2c-ppo-acktr#15]
Great work on the implementation!
Very comprehensible and straightforward implementation.

It seems you're performing two forward steps: 1) to choose an action (main.py, line 113), 2) to evaluate the actions (main.py, line 146).
Why not save the values, log_probs and entropy in while selecting actions (as you did in a3c)?
Are there computational benefits to performing these for all processes at once?

**Comment by ikostrikov** [comment]
In a3c code all computations are performed on CPU. But here I have a GPU option. At some point I compared these 2 strategies and it was faster to compute forward pass for each step and then concatenate, do extra forward pass and then a backward pass on this "large batch".

**Comment by nadavbh12** [comment]
Interesting. It makes sense, though the current implementation makes it more difficult to implement a recurrent policy.
Though I guess whatever way to "batchify" a recurrent policy won't be pleasant in this case (considering that you need to reset the hidden vectors for different envs at different time steps).

**Comment by ikostrikov** [comment]
I agree. It's more difficult to implement a recurrent policy. But it's just marginally more difficult.

I was going to add a recurrent policy in the second half of November anyway. But I can provide you all necessary pointers so you can do it yourself.

**Comment by nadavbh12** [comment]
No need for recurrent at the moment.
I was thinking after n steps to split the matrix in two where the game ends, then use the padded sequence and run them both.
What's your plan?

**Comment by ikostrikov** [comment]
The plan was no reshape them here but send to the model unreshaped. So the FF model reshapes them to batchify and an RNN policy will just unroll it.

https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/blob/master/main.py#L146 

And then we will also need to "duplicate" RNN states. But it should be relatively simple. We just need to store temporary states in first loop and overwrite in the second.

**Comment by nadavbh12** [comment]
Though how do you plan to zero the hidden vector of terminal states?
As I see it you can either:
1) Run the RNN step by step in the "num_steps" loop, zero the hiddens only for processes that reached a terminal state (in this option there's no need for the additional forward afterwards).
2) Split the mini-batch in two where terminal and use padded sequences. As in: Forward first part, zero needed hiddens, forward second part.

I tend towards the first option for clarity and since I don't think you'll have the speed gains you saw earlier with the recurrent policy.

**Comment by ikostrikov** [comment]
I was going to implement it in the following way.

1) save recurrent states before the first loop;
2) unroll to get rewards in a standard way;
3) in the second forward loop retrieve the states saved in 1) and unroll again without reshaping the inputs

This way it will be just easier to reuse the same code for PPO because we can just sample rollouts for different processes in 3)

**Comment by ikostrikov** [comment]
I'm going to add an LSTM policy this week. Do you know a good task to test it? In most papers they use FF policies only.

**Comment by nadavbh12** [comment]
Papers on VisDoom and DeepMind Lab usually use LSTMs, though I haven't tried it myself.

**Comment by ikostrikov** [comment]
Thanks! I will try these games.

At the moment, when I add LSTMs it doesn't even work on the same games.

**Comment by nadavbh12** [comment]
Did you implement an LSTM module for masking the terminal states?
Or some did you do something else?

On Nov 10, 2017 2:06 AM, "Ilya Kostrikov" <notifications@github.com> wrote:

Thanks! I will try these games.

At the moment, when I add LSTMs it doesn't even work on the same games.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/issues/15#issuecomment-343332798>,
or mute the thread

**Comment by ikostrikov** [comment]
Yes, I can push it to a new branch.

**Comment by nadavbh12** [comment]
Cool. I'll give it a look when it's up.

**Comment by ikostrikov** [comment]
It's in the lstm branch:
https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/tree/lstm

You can run it with

```bash
python main.py --env-name "PongNoFrameskip-v4" --recurrent-policy
```

**Comment by ikostrikov** [comment]
After fixing the initialization, the recurrent policy seems to work for a2c (working on ppo as well).

See the LSTM branch.

**Comment by nadavbh12** [comment]
Very nice!
What was the problem with the orthogonal initialization? I can't see a diff to the Pytorch version.
Was it the rows < cols case?

**Comment by ikostrikov** [comment]
Yes, it was rows < cols case. It was fixed in September but it's still not in the version installed with conda.

**Comment by nadavbh12** [comment]
Good thing I sent a [pull request](https://github.com/pytorch/pytorch/pull/2708) :-)

**Comment by ethancaballero** [comment]
any timeline on merging lstm branch into master branch?

**Comment by ikostrikov** [comment]
OpenAI recently released their LSTM PPO. I'm going to derive my code from their implementation now, it it will probably take a week.

**Comment by ikostrikov** [comment]
A recurrent policy is in the main branch now.

**Comment by MarcoMeter** [comment]
@ikostrikov 
I've got a question on how LSTM can be applied to a PPO policy.
While letting agents collect data, how do you treat hx and cx outside of the model?

My assumption was that you simply treat hx and cx as experiences and store them along with their relative state, rewards, logits, for each step of a worker collecting data for updating the policy afterwards.

This is my experience buffer:
```
        rewards = np.zeros((self.n_workers, self.worker_steps), dtype=np.float32)
        actions = np.zeros((self.n_workers, self.worker_steps), dtype=np.int32)
        dones = np.zeros((self.n_workers, self.worker_steps), dtype=np.bool)
        vis_obs = np.zeros((self.n_workers, self.worker_steps) + Game.vis_obs_shape, dtype=np.float32)
        vec_obs = np.zeros((self.n_workers, self.worker_steps,) + Game.vec_obs_shape, dtype=np.float32)
        hx = torch.zeros((self.n_workers, self.worker_steps, 512), dtype=torch.float32, device=self.device)
        cx = torch.zeros((self.n_workers, self.worker_steps, 512), dtype=torch.float32, device=self.device)
        neg_log_pis = np.zeros((self.n_workers, self.worker_steps), dtype=np.float32)
        values = np.zeros((self.n_workers, self.worker_steps), dtype=np.float32)
```
